kyuubi-server/src/main/scala/org/apache/kyuubi/server/statestore/jdbc/JDBCStateStore.scala

```diff
@@ -0,0 +1,438 @@
+
+    executeQuery(queryBuilder.toString(), params: _*)
+  }
+
+  override def cleanupMetadataByIdentifier(identifier: String): Unit = {
+    val query = s"DELETE FROM $METADATA_TABLE WHERE identifier = ?"
+    executeQuery(query, identifier)
+  }
+
+  override def cleanupMetadataByAge(maxAge: Long): Unit = {
+    val minEndTime = System.currentTimeMillis() - maxAge
+    val query = s"DELETE FROM $METADATA_TABLE WHERE end_time IS NOT NULL AND end_time < ?"
+    executeQuery(query, minEndTime)
+  }
+
+  private def buildMetadata(resultSet: ResultSet, stateOnly: Boolean): Seq[Metadata] = {
+    try {
+      val metadataList = ListBuffer[Metadata]()
+      while (resultSet.next()) {
+        val identifier = resultSet.getString("identifier")
+        val realUser = resultSet.getString("real_user")
+        val userName = resultSet.getString("user_name")
+        val ipAddress = resultSet.getString("ip_address")
+        val kyuubiInstance = resultSet.getString("kyuubi_instance")
+        val state = resultSet.getString("state")
+        val requestName = resultSet.getString("request_name")
+        val createTime = resultSet.getLong("create_time")
+        val engineType = resultSet.getString("engine_type")
+        val engineId = resultSet.getString("engine_id")
+        val engineName = resultSet.getString("engine_name")
+        val engineUrl = resultSet.getString("engine_url")
+        val engineState = resultSet.getString("engine_state")
+        val engineError = Option(resultSet.getString("engine_error"))
+        val endTime = resultSet.getLong("end_time")
+
+        var resource: String = null
+        var className: String = null
+        var requestConf: Map[String, String] = Map.empty
+        var requestArgs: Seq[String] = Seq.empty
+
+        if (!stateOnly) {
+          resource = resultSet.getString("resource")
+          className = resultSet.getString("class_name")
+          requestConf = string2Map(resultSet.getString("request_conf"))
+          requestArgs = string2Seq(resultSet.getString("request_args"))
+        }
+        val metadata = Metadata(
+          identifier = identifier,
+          realUser = realUser,
+          username = userName,
+          ipAddress = ipAddress,
+          kyuubiInstance = kyuubiInstance,
+          state = state,
+          resource = resource,
+          className = className,
+          requestName = requestName,
+          requestConf = requestConf,
+          requestArgs = requestArgs,
+          createTime = createTime,
+          engineType = engineType,
+          engineId = engineId,
+          engineName = engineName,
+          engineUrl = engineUrl,
+          engineState = engineState,
+          engineError = engineError,
+          endTime = endTime)
+        metadataList += metadata
+      }
+      metadataList
+    } finally {
+      Utils.tryLogNonFatalError(resultSet.close())
+    }
+  }
+
+  private def execute(conn: Connection, sql: String, params: Any*): ResultSet = {
```

Comment:
it's better to use like `withResultSet` and close each object based on the ordering of `ResultSet` `Statement`